presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTime.java

```diff
@@ -67,6 +84,12 @@ public boolean equals(Object obj)
     @Override
     public String toString()
     {
-        return Instant.ofEpochMilli(millisUtc).atZone(ZoneId.of(sessionTimeZoneKey.getId())).format(formatter);
+        if (isDeprecatedTimestamp()) {
+            return Instant.ofEpochMilli(millisUtc).atZone(ZoneId.of(sessionTimeZoneKey.get().getId())).format(formatter);
+        }
+        else {
+            // FIXME this should be exact equivalent of TIMESTAMP to VARCHAR cast
+            return Instant.ofEpochMilli(millisUtc).atZone(ZoneId.of(UTC_KEY.getId())).format(formatter);
```

Comment:
About this over interpretation.

Semantically, TIMESTAMP W/O TZ (as well as TIME W/O TZ) is just set of integers YEAR (1), MONTH (2), DAY (3), HOUR (4), MINUTE (5), SECOND (6), MILLIS (7).
Our representation encodes that fields into single long that can be interpreted in following way: "If you interpret that long as Instance (epoch in millis), than UTC wall time of this instance (1-7) gives you exact value of (1-7) fields.". This kind of encoding is weird but as far as I understand should work perfectly fine, as each value of such long maps to exactly one UTC wall time and each wall time maps to exactly one value of this representation. Also maths on that is easy for fields (4-7) and it's hard for (1-3) but can be easily done using Instance class.

This is why I think it's perfectly fine (in fact is is the only valid way for fields (1-3)) to use this value as Instance in UTC. For fields (4-7), there is choice to make, and we can use Instance logic or simple duration logic (including basic division/modulo operations) as we know that each day contains exactly `1000*60*60*24` milliseconds.

I use both approaches. When working directly on long value I use TimeUnits for converting values (4-7). I use Instance for everything else as we did use that before and I think that having less changes is better in such huge refactor.

I think that LocalTime is used only in some tests at this point, as you helped me with eliminating that from SqlTimeWithTimeZone.toString(), so I would just not add yet another class to understand our codebase (even though it's class closest to our semantic).

Maybe it is worth considering to drop usage of Instance in all TIME/TIMESTAMP W/O TZ contexts. This is huge change though.

Please let me know if you agree with that approach, as it will (should be) visible across all changes.